pkg/kv/kvserver/replica_application_state_machine.go

```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
 			sm.r.handleReadWriteLocalEvalResult(ctx, *cmd.localResult)
 		}
 
-		if higherReproposalsExist := cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex); higherReproposalsExist {
-			// If the command wasn't rejected, we just applied it and no higher
-			// reproposal must exist (since that one may also apply).
+		if higherReproposalsExist := cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex); higherReproposalsExist &&
+			(!cmd.Rejected() || cmd.Rejection == kvserverbase.ProposalRejectionIllegalLeaseIndex) {
+			// If this entry was accepted, there must not be superseding reproposal,
+			// as nothing prevents it from applying as well in the common case.
+			//
+			// If this entry was rejected but only on account of the lease applied
+			// index, then the call to tryReproposeWithNewLeaseIndex[^1] must have
+			// returned an error (or the proposal would not be IsLocal() now; we
+			// unbind on success). But that call, by design, does not an error for a
+			// proposal that is already superseded.
```

Comment:
This paragraph focuses a bit on the "symptoms": what other methods return or not return. It's a bit hard to follow such logic because of the indirection. Is there a way to frame this in a stateless manner, maybe more abstractly - i.e. directly answer the question why a rejected MaxLeaseIndex can't be found superseded? And then, if needed, employ the additional argument on what the other methods return (which is more like a "proof").